packages/core/src/fhir-deduplication/resources/diagnostic-report.ts

```diff
@@ -92,23 +94,64 @@ export function groupSameDiagnosticReports(diagReports: DiagnosticReport[]): {
     const datetime = getDateFromResource(diagReport, "datetime");
     const isPresentedFormPresent = diagReport.presentedForm?.length;
     const isResultPresent = diagReport.result?.length;
-    if (datetime && (isPresentedFormPresent || isResultPresent)) {
-      const key = JSON.stringify({ datetime });
-      fillMaps(
-        diagReportsMap,
-        key,
-        diagReport,
+
+    // If a diagnostic report does not contain any results or doctor's notes, it is useless and can be removed
+    if (!isPresentedFormPresent && !isResultPresent) {
+      danglingReferences.add(createRef(diagReport));
+      continue;
+    }
+
+    const practitionerRefsSet = new Set<string>();
+
+    const getterKeys: string[] = [];
+    const setterKeys: string[] = [];
+
+    diagReport.performer?.forEach(perf => {
+      const ref = perf.reference;
+      if (ref) {
+        if (ref.includes("Practitioner")) {
+          practitionerRefsSet.add(ref);
+          return;
+        }
+      }
+    });
+
+    const practitionerRefs = Array.from(practitionerRefsSet).map(p => ({ practitioner: p }));
+    if (datetime && practitionerRefs.length > 0) {
+      const practitionerAndDateKeys = createKeysFromObjectArray({ datetime }, practitionerRefs);
+      setterKeys.push(...practitionerAndDateKeys);
+      getterKeys.push(...practitionerAndDateKeys);
+    }
+
+    if (datetime && practitionerRefs.length === 0) {
+      const dateKey = JSON.stringify({ datetime });
+      setterKeys.push(dateKey);
+      getterKeys.push(dateKey);
+    }
+
+    if (!datetime) {
+      const idKey = JSON.stringify({ id: diagReport.id });
+      setterKeys.push(idKey);
+      getterKeys.push(idKey);
+    }
```

Comment:
nit: these are if/elseif, right? maybe worth mapping as such?

I see how the current version is easier on the eyes (auto-formatting keeps empty lines), but any discrepancy in the logic might run more than one of these.

Reply:
good point! fixed!